File: CSharp/ILinkai.Weixin.Sample/Examples/FullWorkflowExample.cs
using ILinkai.Weixin.Sdk;
using ILinkai.Weixin.Sdk.Auth;
using ILinkai.Weixin.Sdk.Messaging;
using ILinkai.Weixin.Sdk.Models;
using Microsoft.Extensions.Logging;
using System.Collections.Concurrent;

namespace ILinkai.Weixin.Sample.Examples;

public class FullWorkflowExample : ExampleBase
{
    private readonly LoginExample _loginExample;
    private static readonly ConcurrentDictionary<string, string> _typingTicketCache = new();

    public override string Name => "full";
    public override string Description => "完整工作流示例";
    public override string Usage => "full";

    public FullWorkflowExample(ILoggerFactory loggerFactory) : base(loggerFactory)
    {
        _loginExample = new LoginExample(loggerFactory);
    }

    public override async Task<int> ExecuteAsync(string[] args)
    {
        PrintHeader("完整工作流示例");
        Console.WriteLine("此示例演示完整的登录->监控->回复流程");
        Console.WriteLine();

        var accountStore = new AccountStore();
        var accounts = accountStore.ListAccountIds();

        string accountId;
        if (accounts.Count == 0)
        {
            Console.WriteLine("没有已登录账户，开始登录流程...");
            Console.WriteLine();

            var loginResult = await _loginExample.ExecuteAsync(args);
            if (loginResult != 0) return loginResult;

            accounts = accountStore.ListAccountIds();
            accountId = accounts[^1];
        }
        else
        {
            Console.WriteLine("请选择账户:");
            Console.WriteLine($"  [0] 新登录账户");
            int i = 0;
            for (; i < accounts.Count; i++)
            {
                Console.WriteLine($"  [{i + 1}] {accounts[i]}");
            }
            Console.WriteLine($"  [{i + 1}] 清理账户");
            Console.WriteLine();

            int selection;
            do
            {
                Console.Write($"请输入选择 (0-{accounts.Count + 1}): ");
            } while (!int.TryParse(Console.ReadLine(), out selection) || selection < 0 || selection > accounts.Count + 1);

            if (selection == 0)
            {
                Console.WriteLine();
                var loginResult = await _loginExample.ExecuteAsync(args);
                if (loginResult != 0) return loginResult;

                accounts = accountStore.ListAccountIds();
                accountId = accounts[^1];
            }
            else if (selection > 0 && selection <= accounts.Count)
            {
                accountId = accounts[selection - 1];
            }
            else
            {
                Console.WriteLine();
                Console.WriteLine("请选择要清理的账户:");
                for (i = 0; i < accounts.Count; i++)
                {
                    Console.WriteLine($"  [{i + 1}] {accounts[i]}");
                }
                Console.WriteLine();
                int clearSelection;
                do
                {
                    Console.Write($"请选择要清理的账户 (1-{accounts.Count}): ");
                } while (!int.TryParse(Console.ReadLine(), out clearSelection) || clearSelection < 1 || clearSelection > accounts.Count);
                var accountToClear = accounts[clearSelection - 1];
                accountStore.ClearAccount(accountToClear);
                Console.WriteLine($"已清理账户: {accountToClear}");
                return 0;
            }
        }

        var account = accountStore.ResolveAccount(accountId);

        Console.WriteLine($"使用账户: {accountId}");
        Console.WriteLine();

        // ── 选择回复模式 ──
        var (replyMode, llmService) = ConfigureReplyMode();

        var apiClient = new WeixinApiClient(new WeixinApiClientOptions
        {
            BaseUrl = account.BaseUrl,
            Token = account.Token
        }, LoggerFactory.CreateLogger<WeixinApiClient>());

        var sendService = new MessageSendService(
            apiClient,
            account.CdnBaseUrl,
            LoggerFactory.CreateLogger<MessageSendService>());

        var options = new MessageMonitorOptions
        {
            BaseUrl = account.BaseUrl,
            CdnBaseUrl = account.CdnBaseUrl,
            Token = account.Token,
            AccountId = account.AccountId,
            MediaDir = Path.Combine(Path.GetTempPath(), "ilinkai-weixin", "media")
        };

        using var monitor = new MessageMonitorService(
            options,
            LoggerFactory.CreateLogger<MessageMonitorService>());

        monitor.MessageReceived += async (sender, e) =>
        {
            Console.WriteLine();
            Console.WriteLine("========================================");
            Console.WriteLine($"收到消息: {e.Context.Body}");
            Console.WriteLine($"发送者: {e.Context.From}");

            if (string.IsNullOrEmpty(e.Context.ContextToken))
            {
                Console.WriteLine("⚠ 无 ContextToken，跳过回复");
            }
            else
            {
                try
                {
                    var typingTicket = await GetTypingTicketAsync(apiClient, e.Context.From);
                    await sendService.SendTypingAsync(e.Context.From, typingTicket);
                    string reply;
                    if (replyMode == ReplyMode.Echo)
                    {
                        reply = FormatEchoReply(e.Message);
                    }
                    else
                    {
                        if (string.IsNullOrWhiteSpace(e.Context.Body))
                        {
                            reply = "(收到非文本消息，暂不支持 LLM 处理)";
                        }
                        else
                        {
                            System.Console.WriteLine("   🤖 正在调用大模型...");
                            reply = await llmService!.ChatAsync(e.Context.Body);
                        }
                    }
                    await sendService.SendTextAsync(e.Context.From, reply, e.Context.ContextToken);
                    Console.WriteLine("已自动回复：" + reply);
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"回复失败: {ex.Message}");
                }
            }
            Console.WriteLine("========================================");
        };

        monitor.ErrorOccurred += (sender, e) =>
        {
            Console.WriteLine($"错误: {e.Error.Message}");
        };

        var cts = new CancellationTokenSource();
        Console.CancelKeyPress += (s, e) =>
        {
            e.Cancel = true;
            cts.Cancel();
        };

        Console.WriteLine("开始监控并自动回复消息，按 Ctrl+C 停止...");
        Console.WriteLine();

        try
        {
            await monitor.StartAsync(cts.Token);
        }
        catch (OperationCanceledException)
        {
            Console.WriteLine("已停止");
        }

        return 0;
    }

    private (ReplyMode mode, LlmService? llm) ConfigureReplyMode()
    {
        System.Console.WriteLine();
        System.Console.WriteLine("选择回复模式:");
        System.Console.WriteLine("  [1] Echo 模式 - 回显收到的消息详情");
        System.Console.WriteLine("  [2] LLM 模式 - 调用大模型对话接口回复");
        System.Console.Write("请选择 (1/2, 默认 1): ");

        var input = System.Console.ReadLine()?.Trim();
        if (input == "2")
        {
            var config = new LlmConfig();

            System.Console.Write($"API Base URL (默认 {config.BaseUrl}): ");
            var baseUrl = System.Console.ReadLine()?.Trim();
            if (!string.IsNullOrEmpty(baseUrl)) config.BaseUrl = baseUrl;

            System.Console.Write("API Key: ");
            string? apiKey = Environment.GetEnvironmentVariable("BIG_MODEL_KEY", EnvironmentVariableTarget.User);
            apiKey = string.IsNullOrEmpty(apiKey) ? System.Console.ReadLine()?.Trim() : apiKey;
            if (string.IsNullOrEmpty(apiKey))
            {
                System.Console.WriteLine("API Key 不能为空，回退到 Echo 模式。");
                return (ReplyMode.Echo, null);
            }
            config.ApiKey = apiKey;

            System.Console.Write($"模型名称 (默认 {config.Model}): ");
            var model = System.Console.ReadLine()?.Trim();
            if (!string.IsNullOrEmpty(model)) config.Model = model;

            System.Console.Write("系统提示词 (可选，回车跳过): ");
            var prompt = System.Console.ReadLine()?.Trim();
            if (!string.IsNullOrEmpty(prompt)) config.SystemPrompt = prompt;

            return (ReplyMode.Llm, new LlmService(config));
        }

        return (ReplyMode.Echo, null);
    }

    private string FormatEchoReply(WeixinMessage msg)
    {
        var sb = new System.Text.StringBuilder();
        sb.AppendLine("📋 消息详情:");
        sb.AppendLine($"  消息ID: {msg.MessageId}");
        sb.AppendLine($"  发送者: {msg.FromUserId}");
        sb.AppendLine($"  接收者: {msg.ToUserId}");
        sb.AppendLine($"  时间: {(msg.CreateTimeMs.HasValue ? DateTimeOffset.FromUnixTimeMilliseconds(msg.CreateTimeMs.Value).ToLocalTime().ToString("yyyy-MM-dd HH:mm:ss") : "未知")}");

        if (msg.ItemList != null)
        {
            foreach (var item in msg.ItemList)
            {
                var typeName = item.Type switch
                {
                    (int)MessageItemType.Text => "文本",
                    (int)MessageItemType.Image => "图片",
                    (int)MessageItemType.Voice => "语音",
                    (int)MessageItemType.File => "文件",
                    (int)MessageItemType.Video => "视频",
                    _ => $"未知({item.Type})",
                };
                sb.AppendLine($"  [{typeName}] {GetItemPreview(item)}");
            }
        }

        return sb.ToString().TrimEnd();
    }

    private string GetItemPreview(MessageItem item)
    {
        if (item.Type == (int)MessageItemType.Text)
            return item.TextItem?.Text ?? "";
        if (item.Type == (int)MessageItemType.Image)
            return $"尺寸: {item.ImageItem?.ThumbWidth}x{item.ImageItem?.ThumbHeight}";
        if (item.Type == (int)MessageItemType.Voice)
            return $"时长: {item.VoiceItem?.PlayTime}ms | 转文字: {item.VoiceItem?.Text ?? "无"}";
        if (item.Type == (int)MessageItemType.File)
            return $"{item.FileItem?.FileName} ({item.FileItem?.Length} bytes)";
        if (item.Type == (int)MessageItemType.Video)
            return $"时长: {item.VideoItem?.PlayLength}s";
        return "";
    }

    private async Task<string> GetTypingTicketAsync(WeixinApiClient apiClient, string userId)
    {
        if (_typingTicketCache.TryGetValue(userId, out var cached))
            return cached;

        var config = await apiClient.GetConfigAsync(userId);
        var ticket = config?.TypingTicket ?? string.Empty;
        _typingTicketCache[userId] = ticket;
        return ticket;
    }

    enum ReplyMode { Echo, Llm }
}

File: Python/samples/full_workflow_example.py
"""完整工作流示例"""
import os
import tempfile
from datetime import datetime

from samples.example_base import ExampleBase
from samples.login_example import LoginExample
from samples.llm_service import LlmService, LlmConfig
from ilinkai_weixin import WeixinApiClient, WeixinApiClientOptions
from ilinkai_weixin.auth.account_store import AccountStore
from ilinkai_weixin.messaging.message_send_service import MessageSendService
from ilinkai_weixin.messaging.message_monitor_service import MessageMonitorService
from ilinkai_weixin.models import MessageItemType


class FullWorkflowExample(ExampleBase):
    name = "full"
    description = "完整工作流示例"
    usage = "full"

    def __init__(self):
        self._login_example = LoginExample()

    def execute(self, args: list) -> int:
        self.print_header("完整工作流示例")
        print("此示例演示完整的登录->监控->回复流程")
        print()

        account_store = AccountStore()
        accounts = account_store.list_account_ids()

        if not accounts:
            print("没有已登录账户，开始登录流程...")
            print()
            login_result = self._login_example.execute(args)
            if login_result != 0:
                return login_result
            accounts = account_store.list_account_ids()
            account_id = accounts[-1]
        else:
            print("请选择账户:")
            print("  [0] 新登录账户")
            for i, acc in enumerate(accounts):
                print(f"  [{i + 1}] {acc}")
            print()

            selection = -1
            while selection < 0 or selection > len(accounts):
                try:
                    selection = int(input(f"请输入选择 (0-{len(accounts)}): ").strip())
                except (ValueError, EOFError):
                    selection = -1

            if selection == 0:
                print()
                login_result = self._login_example.execute(args)
                if login_result != 0:
                    return login_result
                accounts = account_store.list_account_ids()
                account_id = accounts[-1]
            else:
                account_id = accounts[selection - 1]

        account = account_store.resolve_account(account_id)
        print(f"使用账户: {account_id}")
        print()

        # 选择回复模式
        use_llm = False
        llm_service = None

        print("选择回复模式:")
        print("  [1] Echo 模式 - 回显收到的消息详情")
        print("  [2] LLM 模式 - 调用大模型对话接口回复")
        mode_input = input("请选择 (1/2, 默认 1): ").strip()

        if mode_input == "2":
            config = LlmConfig()
            base_url = input(f"API Base URL (默认 {config.base_url}): ").strip()
            if base_url:
                config.base_url = base_url

            api_key = input("API Key: ").strip()
            if not api_key:
                print("API Key 不能为空，回退到 Echo 模式。")
            else:
                config.api_key = api_key
                model = input(f"模型名称 (默认 {config.model}): ").strip()
                if model:
                    config.model = model
                prompt = input("系统提示词 (可选，回车跳过): ").strip()
                if prompt:
                    config.system_prompt = prompt
                llm_service = LlmService(config)
                use_llm = True

        client = WeixinApiClient(WeixinApiClientOptions(
            base_url=account.base_url, token=account.token))
        send_service = MessageSendService(client, account.cdn_base_url)

        media_dir = os.path.join(tempfile.gettempdir(), "ilinkai-weixin", "media")
        monitor = MessageMonitorService(
            base_url=account.base_url, cdn_base_url=account.cdn_base_url,
            token=account.token, account_id=account.account_id, media_dir=media_dir)

        def on_message(ctx):
            print()
            print("========================================")
            print(f"收到消息: {ctx.body}")
            print(f"发送者: {ctx.from_}")

            if not ctx.context_token:
                print("⚠ 无 ContextToken，跳过回复")
            else:
                try:
                    if use_llm and llm_service:
                        if not ctx.body or not ctx.body.strip():
                            reply = "(收到非文本消息，暂不支持 LLM 处理)"
                        else:
                            print("   🤖 正在调用大模型...")
                            reply = llm_service.chat(ctx.body)
                    else:
                        reply = _format_echo_reply(ctx.original_message)
                    send_service.send_text(ctx.from_, reply, ctx.context_token)
                    print("已自动回复：" + reply)
                except Exception as ex:
                    print(f"回复失败: {ex}")
            print("========================================")

        def on_error(err):
            print(f"错误: {err}")

        monitor.on_message_received = on_message
        monitor.on_error = on_error

        print("开始监控并自动回复消息，按 Ctrl+C 停止...")
        print()

        try:
            monitor.start()
        except KeyboardInterrupt:
            print("\n已停止")
        return 0


def _format_echo_reply(msg):
    if not msg:
        return "(空消息)"
    lines = ["📋 消息详情:"]
    lines.append(f"  消息ID: {msg.message_id}")
    lines.append(f"  发送者: {msg.from_user_id}")
    lines.append(f"  接收者: {msg.to_user_id}")
    if msg.create_time_ms:
        time_str = datetime.fromtimestamp(msg.create_time_ms / 1000).strftime("%Y-%m-%d %H:%M:%S")
        lines.append(f"  时间: {time_str}")
    if msg.item_list:
        for item in msg.item_list:
            t = item.type or 0
            type_names = {
                MessageItemType.TEXT: "文本", MessageItemType.IMAGE: "图片",
                MessageItemType.VOICE: "语音", MessageItemType.FILE: "文件",
                MessageItemType.VIDEO: "视频",
            }
            type_name = type_names.get(t, f"未知({t})")
            preview = _get_item_preview(item)
            lines.append(f"  [{type_name}] {preview}")
    return "\n".join(lines)


def _get_item_preview(item):
    t = item.type or 0
    if t == MessageItemType.TEXT:
        return item.text_item.text if item.text_item else ""
    if t == MessageItemType.IMAGE and item.image_item:
        return f"尺寸: {item.image_item.thumb_width}x{item.image_item.thumb_height}"
    if t == MessageItemType.VOICE and item.voice_item:
        return f"时长: {item.voice_item.play_time}ms | 转文字: {item.voice_item.text or '无'}"
    if t == MessageItemType.FILE and item.file_item:
        return f"{item.file_item.file_name} ({item.file_item.length} bytes)"
    if t == MessageItemType.VIDEO and item.video_item:
        return f"时长: {item.video_item.play_length}s"
    return ""
